### PR TITLE
Update EIP-2315 implementation (enter only via jumpsub)

### PR DESCRIPTION
**Pull request by matkt**
Signed-off-by: Karim TAAM <karim.t2am@gmail.com>

## PR description

Update implementation of the EIP-2315 after the recent PR (https://github.com/ethereum/EIPs/pull/2646)

- Execution of BEGINSUB causes exception (OOG: all gas consumed) and terminates execution. 
- JUMPSUB sets the pc to location + 1 .
- In the edge case when BEGINSUB is the last instruction in code and this subroutine is jumped-to, the implementations should execute STOP .

Tested with https://github.com/ethereum/tests/pull/685

### CODE REVIEW

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/BerlinGasCalculator.java [hunk 1]
```diff
@@ -18,7 +18,7 @@
 
 public class BerlinGasCalculator extends IstanbulGasCalculator {
 
-  private static final Gas BEGIN_SUB_GAS_COST = Gas.of(1);
+  private static final Gas BEGIN_SUB_GAS_COST = Gas.of(0);
```
Comment (RatanRSur) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/BerlinGasCalculator.java:21]: (optional) `Gas.ZERO`

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/BeginSubOperationTest.java [hunk 2]
```diff
@@ -42,7 +43,7 @@
   private static final GasCalculator gasCalculator = new BerlinGasCalculator();
 
   private static final int CURRENT_PC = 1;
-  private static final Gas BEGIN_SUB_GAS_COST = Gas.of(1);
+  private static final Gas BEGIN_SUB_GAS_COST = Gas.of(0);
```
Comment (RatanRSur) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/BeginSubOperationTest.java:46]: same as above

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/BeginSubOperationTest.java [hunk 3]
```diff
@@ -85,17 +86,16 @@ public void shouldCalculateGasPrice() {
...
-  public void shouldNotConsumeStackItem() {
+  public void shouldHaltWithOutOfGasWhenBeginSubIsExecuted() {
     final BeginSubOperation operation = new BeginSubOperation(gasCalculator);
     final MessageFrame frame =
         createMessageFrameBuilder(Gas.of(1))
-            .pushStackItem(Bytes32.fromHexString("0xFF"))
-            .code(new Code(Bytes.fromHexString("0x6004b300b2b7")))
+            .pushStackItem(Bytes32.fromHexString("0x04"))
+            .code(new Code(Bytes.fromHexString("0x610400b2")))
             .returnStack(new ReturnStack(MessageFrame.DEFAULT_MAX_RETURN_STACK_SIZE))
```
Comment (RatanRSur) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/BeginSubOperationTest.java:95]: I guess this is a comment for another PR but why do we have the `ReturnStack` parameterized by size? The size of it is fixed in the EIP, yes? So we shouldn't expose that detail to the caller.

Comment (matkt) [reply ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/BeginSubOperationTest.java:95]: yes you're right I always make this call `new ReturnStack (MessageFrame.DEFAULT_MAX_RETURN_STACK_SIZE)`  I will set this value on the ReturnStack object. I will do another PR to fix that

Comment (RatanRSur) [reply ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/BeginSubOperationTest.java:95]: Cool, then you can model it as a singleton class with a static `ReturnStack`, and private constructor.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/BerlinGasCalculator.java (hunk 1 above)
Comment (matkt) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/BerlinGasCalculator.java]: Done

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/BeginSubOperationTest.java (hunk 2 above)
Comment (matkt) [reply ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/BeginSubOperationTest.java]: Done

**Pull request merged**